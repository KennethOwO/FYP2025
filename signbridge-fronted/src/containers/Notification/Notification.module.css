.notifocationBigContainer {
    padding: 20px;
    height: 100%;
    margin-bottom: 50px;
}

.notifocationWholeContainer {
    display: flex;
    justify-content: center;
    margin: 20px 0 20px 0;
}

.notifocationHeader {
    text-align: center;
}

.notifocationHeaderText {
    font-size: 30px;
    color: rgb(0, 0, 0);
    margin: 30px;
}

.notificationContainer {
    justify-items: start;
    text-align: center;
    width: 60%;
    margin-left: 80px;
}

.toolsItem {
    display: flex;
    align-items: center;
    width: 100%;
    background-color: #f6f7f9;
    border-radius: 5px 5px 0 0;
}

.toolsIcon {
    cursor: pointer;
    border-radius: 50%;
    padding: 10px;
}

.toolsIcon:hover {
    background-color: #edf2f7;
}

.noNotificationsContainer {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    background-color: #f6f7f9;
    height: 80%;
    width: 100%;
}

.noNotificationsContainer p {
    font-weight: 400;
    font-size: 16px;
    color: rgb(128, 129, 145);
    text-align: center;
}

@media screen and (max-width: 768px) {
    .notifocationBigContainer {
        padding: 20px;
    }

    .notificationContainer {
        flex-direction: column;
        gap: 20px;
        margin-left: 20px;
    }
}

@media screen and (max-width: 580px) {
    .notifocationWholeContainer {
        flex-direction: column;
        gap: 20px;
        align-items: center;
    }

    .notificationContainer {
        width: 90%;
        margin-left: 0;
    }

    .toolsIcon {
        padding: 8px;
    }

    .notifocationHeaderText {
        font-size: 20px;
    }

    .noNotificationsContainer{
        padding: 20px;
    }

    .noNotificationsContainer p {
        font-size: 14px;
    }
}
