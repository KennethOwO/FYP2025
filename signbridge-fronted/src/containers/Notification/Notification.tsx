import React, { useEffect, useState } from "react";
import style from "./Notification.module.css";
import NotifBox from "../../components/NotifBox/NotifBox";
import NotifFilter from "../../components/NotifFilter/NotifFilter";
import { useNotificationFilterStore } from "../../store/notificationFilter";
import Checkbox from "@mui/material/Checkbox";
import Button from "@mui/material/Button";
import Tooltip from "@mui/material/Tooltip";
import { toast } from "react-hot-toast";
import { GetSenderInfoBySenderId, DeleteNotification, UpdateNotificationStatus, FetchNotificationsDataByStream } from "../../services/notification.service";
import { useTranslation } from "react-i18next";
import { useUserStore } from "@root/store/userStore";
import { getAuth } from "firebase/auth";

let isUndoing = false;

const Notification: React.FC = () => {
    const { t } = useTranslation();
    const currentSelectedLanguage = localStorage.getItem("i18nextLng");
    const { user } = useUserStore();
    const useFilterStore = useNotificationFilterStore();
    const currentUser = getAuth().currentUser;
    const [selectedNotifications, setSelectedNotifications] = useState<number[]>([]);
    const [timerId, setTimerId] = useState<NodeJS.Timeout | null>(null);

    const filterData = () => {
        let newData: any[] = [];

        if (useFilterStore.filter.includes("newtask")) {
            // add the data that has newtask value
            for (let i = 0; i < useFilterStore.data.length; i++) {
                if (useFilterStore.data[i].type_value === "newtask") {
                    newData.push(useFilterStore.data[i]);
                }
            }
        }

        if (useFilterStore.filter.includes("accepted")) {
            // add the data that has accepted value, it should be added to the new data instead of replacing it
            for (let i = 0; i < useFilterStore.data.length; i++) {
                if (useFilterStore.data[i].type_value === "accepted") {
                    newData.push(useFilterStore.data[i]);
                }
            }
        }

        if (useFilterStore.filter.includes("rejected")) {
            // add the data that has rejected value, it should be added to the new data instead of replacing it
            for (let i = 0; i < useFilterStore.data.length; i++) {
                if (useFilterStore.data[i].type_value === "rejected") {
                    newData.push(useFilterStore.data[i]);
                }
            }
        }

        if (useFilterStore.filter.includes("newtext")) {
            // add the data that has newtext value
            for (let i = 0; i < useFilterStore.data.length; i++) {
                if (useFilterStore.data[i].type_value === "newtext") {
                    newData.push(useFilterStore.data[i]);
                }
            }
        }

        if (useFilterStore.filter.includes("waitingforverification")) {
            // add the data that has waitingforverification value
            for (let i = 0; i < useFilterStore.data.length; i++) {
                if (useFilterStore.data[i].type_value === "waitingforverification") {
                    newData.push(useFilterStore.data[i]);
                }
            }
        }

        newData = newData.sort((a, b) => new Date(b.created_at).getTime() - new Date(a.created_at).getTime());
        useFilterStore.setModifiedData(newData);
    };

    useEffect(() => {
        let eventSourceCleanup: (() => void) | undefined;

        const setupNotificationStream = async () => {
            if (isUndoing || !user || !currentUser) return;

            try {
                eventSourceCleanup = await FetchNotificationsDataByStream(user.user_id, async notifications => {
                    const notificationsWithUsernames = await Promise.all(
                        notifications.map(async (notification: any) => {
                            const senderInfo = await GetSenderInfoBySenderId(notification.sender_id, currentUser);
                            return {
                                ...notification,
                                sender_username: senderInfo.data.username,
                                sender_avatar: senderInfo.data.picture,
                            };
                        })
                    );

                    const data = notificationsWithUsernames.sort((a, b) => new Date(b.created_at).getTime() - new Date(a.created_at).getTime());
                    useFilterStore.setData(data);
                });
            } catch (error) {
                console.error("Error in notification stream:", error);
                // Attempt to reconnect after a delay
                setTimeout(() => {
                    setupNotificationStream();
                }, 5000);
            }
        };

        setupNotificationStream();

        return () => {
            if (eventSourceCleanup) {
                eventSourceCleanup();
            }
        };
    }, [user, currentUser, isUndoing]);

    useEffect(() => {
        filterData();
    }, [useFilterStore.data]);

    const formatDate = (date: string) => {
        const today = new Date();
        const newDate = new Date(date);
        if (today.getDate() === newDate.getDate() && today.getMonth() === newDate.getMonth() && today.getFullYear() === newDate.getFullYear()) {
            return new Date(date).toLocaleTimeString();
        } else {
            return new Date(date).toLocaleDateString();
        }
    };

    const handleSelectAll = (e: React.ChangeEvent<HTMLInputElement>) => {
        if (e.target.checked) {
            // Select all notifications
            const allNotificationIds = useFilterStore.modifiedData.map(notification => notification.notification_id);
            setSelectedNotifications(allNotificationIds);
        } else {
            // Deselect all notifications
            setSelectedNotifications([]);
        }
    };

    // handle the delete icon function
    const handleDeleteNotification = async (notificationIds: number[]) => {
        isUndoing = true;
        handleDelete(notificationIds);
        try {
            const originalData = [...useFilterStore.modifiedData];
            const updatedData = useFilterStore.modifiedData.filter(notification => !notificationIds.includes(notification.notification_id));
            useFilterStore.setModifiedData(updatedData);

            toast.success(s => (
                <span>
                    {t("notifDeleteSuccess")}
                    <Button
                        onClick={() => {
                            useFilterStore.setModifiedData(originalData);
                            handleUndo();
                            toast.dismiss(s.id);
                        }}>
                        UNDO
                    </Button>
                </span>
            ));
        } catch (error) {
            console.error("Error deleting notification:", error);
            toast.error(t("notifDeleteFailed"));
        }
    };

    const handleDelete = async (notificationIds: number[]) => {
        if (!timerId && currentUser) {
            const id = setTimeout(async () => {
                await DeleteNotification(notificationIds, currentUser);
                setTimerId(null);
                isUndoing = false;
            }, 5000);
            setTimerId(id);
        }
    };

    const handleUndo = () => {
        if (timerId) {
            clearTimeout(timerId);
            setTimerId(null);
            isUndoing = false;
        }
    };

    // handle the make as read icon function
    const handleMakeAsRead = async (notificationIds: number[]) => {
        try {
            if (currentUser) {
                await UpdateNotificationStatus(notificationIds, 1, currentUser);
                toast.success(t("notifReadSuccess"));
            }
        } catch (error) {
            console.error("Error updating notification status:", error);
            toast.error(t("notifReadFailed"));
        }
    };

    return (
        <>
            <div className={style.notifocationBigContainer}>
                <div className={style.notifocationHeader}>
                    <h1 className={style.notifocationHeaderText}>{t("notification")}</h1>
                </div>
                <div className={style.notifocationWholeContainer}>
                    <NotifFilter />
                    <div className={style.notificationContainer}>
                        {/* Tools container */}
                        <div className={style.toolsContainer}>
                            <div className={style.toolsItem}>
                                <Tooltip title={t("selectAll")} arrow followCursor>
                                    <Checkbox checked={selectedNotifications.length === useFilterStore.modifiedData.length && useFilterStore.modifiedData.length > 0} indeterminate={selectedNotifications.length > 0 && selectedNotifications.length < useFilterStore.modifiedData.length} onChange={handleSelectAll} />
                                </Tooltip>
                                <div className={style.toolsIcon}>
                                    <Tooltip title={t("delete")} arrow followCursor>
                                        <i className="fa-regular fa-trash-can" onClick={() => handleDeleteNotification(selectedNotifications)}></i>
                                    </Tooltip>
                                </div>
                                <div className={style.toolsIcon}>
                                    <Tooltip title={t("markAsRead")} arrow followCursor>
                                        <i className="fa-regular fa-envelope-open" onClick={() => handleMakeAsRead(selectedNotifications)}></i>
                                    </Tooltip>
                                </div>
                            </div>
                        </div>
                        {useFilterStore.modifiedData.length > 0 ? (
                            useFilterStore.modifiedData.map((notification, index) => (
                                <NotifBox
                                    key={index}
                                    sender_username={notification.sender_username}
                                    sender_avatar={notification.sender_avatar}
                                    message={currentSelectedLanguage === "en" ? notification.message_en : notification.message_bm}
                                    created_at={formatDate(notification.created_at)}
                                    sign_text={notification.sign_text}
                                    status={notification.status}
                                    checked={selectedNotifications.includes(notification.notification_id)}
                                    handleCheckboxChange={e => {
                                        if (e.target.checked) {
                                            setSelectedNotifications([...selectedNotifications, notification.notification_id]);
                                        } else {
                                            setSelectedNotifications(selectedNotifications.filter(id => id !== notification.notification_id));
                                        }
                                    }}
                                />
                            ))
                        ) : (
                            <div className={style.noNotificationsContainer}>
                                <p>{t("noNotif")}</p>
                            </div>
                        )}
                    </div>
                </div>
            </div>
        </>
    );
};

export default Notification;
