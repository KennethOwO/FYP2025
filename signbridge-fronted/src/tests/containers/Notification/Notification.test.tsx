import React from "react";
import { render, cleanup } from "@testing-library/react";
import Notification from "@root/containers/Notification/Notification";

jest.mock("react-i18next", () => ({
  useTranslation: () => ({
    t: (key: string) => key,
    i18n: {
      changeLanguage: jest.fn(),
    },
  }),
}));

describe("Test Notification", () => {
  afterEach(() => {
    cleanup();
    jest.resetAllMocks();
  });

  it("should render correctly", () => {
    const { container } = render(<Notification />);

    expect(container).toMatchSnapshot();
  });
});
