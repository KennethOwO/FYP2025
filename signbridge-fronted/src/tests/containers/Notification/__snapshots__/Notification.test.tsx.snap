// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Notification should render correctly 1`] = `
<div>
  <div>
    <div>
      <h1>
        notification
      </h1>
    </div>
    <div>
      <div>
        <div>
          <p>
            filter
          </p>
        </div>
        <div>
          <div
            class="MuiFormGroup-root css-dmmspl-MuiFormGroup-root"
          />
        </div>
      </div>
      <div>
        <div>
          <div>
            <span
              aria-label="selectAll"
              class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
              data-mui-internal-clone-element="true"
            >
              <input
                class="PrivateSwitchBase-input css-1m9pwf3"
                data-indeterminate="false"
                type="checkbox"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="CheckBoxOutlineBlankIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </span>
            <div>
              <i
                aria-label="delete"
                class="fa-regular fa-trash-can"
                data-mui-internal-clone-element="true"
              />
            </div>
            <div>
              <i
                aria-label="markAsRead"
                class="fa-regular fa-envelope-open"
                data-mui-internal-clone-element="true"
              />
            </div>
          </div>
        </div>
        <div>
          <p>
            noNotif
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
